csrc/multidevice/communication.cpp

```diff
@@ -111,209 +98,225 @@ Broadcast::Broadcast(CommParams params) : Communication(params, "broadcast") {}
 
 c10::intrusive_ptr<c10d::Work> Broadcast::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
 
   if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.src_bufs, 1);
-    if (params_.dst_bufs.size() == 1) {
-      doLocalCopy(params_.dst_bufs.at(0), params_.src_bufs.at(0));
+    if (params_.is_root_in_mesh) {
+      // Do a local copy and the subsequent broadcast will be in place.
```

Comment:
I don't think I've changed the behavior, but correct me if I'm wrong. 

> I think we shouldn't handle the copy ourselves in this case

I'm happy to learn how to avoid the copy with the current c10d::Backend::broadcast API. Am I missing a broadcast_oop somewhere?

Reply:
> Nice! Is [in-place broadcast](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/colls.html#ncclbroadcast) still preferred to out-of-place when we don't need to local copy?

As far as NCCL is concerned, the answer is "no" according to that comment saying that [inplace bcast is deprecated](https://github.com/NVIDIA/nccl/blob/ab2b89c4c339bd7f816fbc114a4b05d386b66290/src/collectives.cc#L75) (because it duplicates functionality). But in practice in shouldn't make any difference.

For UCC, there is no such an "out-of-place broadcast" (because this primitive doesn't exists in MPI standard), so we'll need to stick to (inplace) broadcast anyway.

> Currently, input_tensor and output_tensor have different buffers, but we can change nvFuser to make them alias by calling `Fusion::aliasOutputToInput`, if the input is not used elsewhere e.g. saved for backprop.

I see, yes that could be very interesting.
BTw, I'd love to know more about the status of Aliasing. Can we Alias any two given intermediate tensors, or does it need to be Inputs and outputs? also, is there an IR symbolizing aliasing relationship, or do we plan to add such an IR?


Im resolving this conversation since it is not related to the pr anymore